### ISSUE TITLE
Python install has no recommended package... 

### ISSUE THREAD

**Issue by jgeewax** [grpc/grpc#4118]
Related to #4109 

```sh
$ sudo apt-get install libgrpc-dev python-dev python-pip
$ sudo pip install grpcio
Downloading/unpacking grpcio
  Could not find a version that satisfies the requirement grpcio (from versions: 0.10.0a0, 0.11.0b0, 0.11.0b1, 0.4.0a0, 0.4.0a1, 0.4.0a13, 0.4.0a14, 0.4.0a2, 0.4.0a3, 0.4.0a4, 0.4.0a5, 0.4.0a6, 0.4.0a7, 0.4.0a8, 0.5.0a0, 0.5.0a1, 0.5.0a2, 0.9.0a0, 0.9.0a1)
Cleaning up...
No distributions matching the version for grpcio
Storing debug log for failure in /root/.pip/pip.log
```

The natural thing to do is look at the end of the list... Things count up, right? Wrong! The highest version is actually 0.11.0b1!

When I do take the one from the end of the list (0.9.0a1), I get lots of errors.

Further, I think most people are going to be like "Wait, wtf? I can't just `pip install grpcio` ?"

Can we make something a default version on PyPI so that people don't have to figure out this `==?` stuff?

---

Just FYI, here's what happens when I install the `libgrpc-dev` package and then try to use `grpcio==0.9.0a1` :

```sh
$ sudo apt-get install libgrpc-dev python-dev python-pip
$ sudo pip install grpcio==0.9.0a1
... snip ...
       from grpc/_adapter/_call.c:34:

/usr/include/grpc/grpc.h:525:17: note: declared here

 grpc_call_error grpc_call_start_batch(grpc_call *call, const grpc_op *ops,

                 ^

grpc/_adapter/_call.c: In function ‘pygrpc_call_status’:

grpc/_adapter/_call.c:354:16: error: too few arguments to function ‘grpc_call_start_batch’

   call_error = grpc_call_start_batch(self->c_call, &op, 1, c_tag);

                ^

In file included from ./grpc/_adapter/_call.h:38:0,

                 from grpc/_adapter/_call.c:34:

/usr/include/grpc/grpc.h:525:17: note: declared here

 grpc_call_error grpc_call_start_batch(grpc_call *call, const grpc_op *ops,

                 ^

grpc/_adapter/_call.c: In function ‘pygrpc_call_cancel’:

grpc/_adapter/_call.c:363:38: error: too few arguments to function ‘grpc_call_cancel’

   return pygrpc_translate_call_error(grpc_call_cancel(self->c_call));

                                      ^

In file included from ./grpc/_adapter/_call.h:38:0,

                 from grpc/_adapter/_call.c:34:

/usr/include/grpc/grpc.h:580:17: note: declared here

 grpc_call_error grpc_call_cancel(grpc_call *call, void *reserved);

                 ^

grpc/_adapter/_call.c:364:1: warning: control reaches end of non-void function [-Wreturn-type]

 }

 ^

error: command 'x86_64-linux-gnu-gcc' failed with exit status 1

----------------------------------------
Cleaning up...
Command /usr/bin/python -c "import setuptools, tokenize;__file__='/tmp/pip-build-TGyR1A/grpcio/setup.py';exec(compile(getattr(tokenize, 'open', open)(__file__).read().replace('\r\n', '\n'), __file__, 'exec'))" install --record /tmp/pip-eNXxu2-record/install-record.txt --single-version-externally-managed --compile failed with error code 1 in /tmp/pip-build-TGyR1A/grpcio
Traceback (most recent call last):
  File "/usr/bin/pip", line 9, in <module>
    load_entry_point('pip==1.5.6', 'console_scripts', 'pip')()
  File "/usr/lib/python2.7/dist-packages/pip/__init__.py", line 248, in main
    return command.main(cmd_args)
  File "/usr/lib/python2.7/dist-packages/pip/basecommand.py", line 161, in main
    text = '\n'.join(complete_log)
UnicodeDecodeError: 'ascii' codec can't decode byte 0xe2 in position 39: ordinal not in range(128)
```

**Comment by soltanmm** [comment]
I'm not sure what you mean by 'things count up' --> 'wrong'. Why is 0.9.x considered more towards the 'end' than 0.11.x?

**Comment by jgeewax** [comment]
It looks like PIP does alphabetical sorting when showing you the list of available options, but our versions count numerically for each "group" (separated by dots):

For example, let's say we have versions: `0.1, 0.9, 0.11, 1.0` (for the illustration, I'm ignoring non-numeric characters)

PIP sorted (note that it's just saying "let's sort the numbers")

```python
>>> versions = ['0.1', '0.9', '0.11', '1.0', '0.5']
>>> sorted(versions)
['0.1', '0.11', '0.5', '0.9', '1.0']
```

In other words, sorting this way treats `0.5

"What a human actually wants" (which compares each number between the dots)
```python
>>> def compare_versions(a, b):
...     pieces_a, pieces_b = a.split('.'), b.split('.')
...     pieces_a, pieces_b = map(float, pieces_a), map(float, pieces_b)
... 
...     # Make sure they're the right length
...     length_diff = len(pieces_a) - len(pieces_b)
...     if length_diff > 0: pieces_b += [0]*length_diff
...     elif length_diff < 0: pieces_a += [0]*-length_diff
... 
...     for (piece_a, piece_b) in zip(pieces_a, pieces_b):
...         if piece_a < piece_b: return -1
...         elif piece_a > piece_b: return 1
... 
...     return 0
>>> 
>>> print sorted(versions, compare_versions)
['0.1', '0.9', '0.11', '1.0']
```

**Comment by jgeewax** [comment]
Forgive the long comment -- I got all geeky on this...

It looks like PIP does alphabetical sorting rather than using `distutils.version` sorting. This shows up as (not rendering as code here so you can see everything without scrolling):

$ sudo pip install grpcio
Downloading/unpacking grpcio
  Could not find a version that satisfies the requirement grpcio (from versions: 0.10.0a0, 0.11.0b0, 0.11.0b1, 0.4.0a0, 0.4.0a1, 0.4.0a13, 0.4.0a14, 0.4.0a2, 0.4.0a3, 0.4.0a4, 0.4.0a5, 0.4.0a6, 0.4.0a7, 0.4.0a8, 0.5.0a0, 0.5.0a1, 0.5.0a2, 0.9.0a0, 0.9.0a1)
Cleaning up...

Note that the last one in the list is `0.9.0a1`, but the "most recent" (if PIP was sorting based on `distutils.version`) would be `0.11.0b1`. I figure the last one is the highest number (I see it's certainly bigger than `0.5.0a2`...) so I just go with that, which then leads me to compile errors.

**We can't fix how PIP does things in the scope of this issue, but we can set a default version so that this choice isn't required...**

---

If you're interested, it looks like PIP is just calling `sorted` on the versions when showing you the list of available options, but that isn't really how versions should be sorted.

For example, let's say we have versions: `0.1, 0.9, 0.11b1, 1.0`

PIP sorted (note that it's just saying "let's sort the strings")

```python
>>> versions = ['0.1', '0.9', '0.11b1', '1.0', '0.5']
>>> sorted(versions)
['0.1', '0.11b1', '0.5', '0.9', '1.0']
```

What we expect here is what `distutils.version` provides, where we're treating each "section" (separated by dots) as it's own piece:

```python
>>> from distutils.version import StrictVersion
>>> versions = ['0.1', '0.9', '0.11', '1.0', '0.5']
>>> print map(str, sorted(map(StrictVersion, versions)))
['0.1', '0.5', '0.9', '0.11b1', '1.0']
```

In other words, PIP is sorting in a way that treats version `0.5` and `0.50` as the same thing, whereas most versions treat `0.5` and `0.50` numerically as comparing `0.05` and `0.50` (and read it as "oh dot five" and "oh dot fifty")

**Comment by soltanmm** [comment]
I think we 'solved' (as best one might) this a while back by hiding older versions by default on PyPI. Feel free to re-open if my claim is bonkers.
